add feature: support for managing groups by project

JeanTracker opened the pull request:
Add a menu for project management to the Activity Bar. Users can select a project
from the project list, and the filter groups for the selected project will be loaded.
By default, filter and group properties are loaded as disabled.

Signed-off-by: JeanTracker <hyojoong.kim.jean@gmail.com>

JeanTracker commented:
This is a PR for second feature of #7 . This PR should be merged after #14.

JeanTracker commented:
I got the idea for this from the other extension [projects+](https://marketplace.visualstudio.com/items?itemName=fabiospampinato.vscode-projects-plus). The core concept is to require project creation, and within each project, you can create and manage filter groups. I removed the import/export functionality and instead manage the project file directly. Currently, the settings are recorded in a `vscode_log_filter.json` file in the home directory, but I believe adding extension settings for flexibility would be beneficial in the future.

SoySauceFor3 commented:
We tried this PR in debugger. we could not get the extension to work (explained later), but we have some questions about this PR. 

1. The benefit of this project feature: 
   If my deduction about how this feature works is correct, it seems to me the main benefit is to have groups of filters saved for workspaces (because if it is for single folder, a saved json file for groups is enough). Is this correct? If so, it might not be that big a benefit. Also, even if it is a big benefit, I think the UX needs more consideration, as right now it looks like the user has to use their brain to pair "project" with "workspace". 
   Pls let us know your main purpose of this feature, and provide some reasoning behind the trade off between the benefit and the added complexity. 
2. If projects are mandatory, they add a hurdle for users who don't need the additional organizations of projects. It's good if there is a clear and simple path for users who may not need that. Forcing users into a structure they don't require could lead to frustration and reduced adoption.
3. If the previous concerns are resolved, and we are keeping this "project" feature: the current UX you have has parts separated in two tabs in the activity bar. It can be quite confusing and not natural. It’s important that the workflow feels intuitive and natural for users, minimizing cognitive load and avoiding unnecessary steps.
4. Lastly, in regards to implementation: when we test this PR in debugger, we added two projects in the "Log Analysis" tab, but when we try to add a group in the "FILTER" drop down, it shows error msg saying we have to add a project first. So there might be some bugs in the implementation. 

We appreciate the effort you've put into this feature and look forward to continue the discussion. Please let us know if you need any further clarification on our feedback or if you would like to discuss these points in more detail.

JeanTracker commented:
The project feature was added to enhance log analysis. I initially wanted to discuss this aspect, but since you mentioned that the PR review format is more convenient, I shared the development progress first. I understand that from a UX perspective, it may seem a bit incomplete at this stage.

To explain the background behind introducing the project feature

Currently, user can create filter lists for log analysis, but these lists disappear when user close vscode. Users are responsible for managing these filter lists, and while import/export functionality is available, it becomes inconvenient when dealing with multiple lists. Specifically, switching between filter lists is inconvenient because import only adds to the current list, requiring a vscode restart to reconfigure the filters properly. This might be fine for dedicated log analysis tools like textAnalysisTool.NET, but since vscode is a multi-purpose tool used for various tasks like coding, having to restart it just for log analysis is a significant week point.

This led to the introduction of the project feature, with the main goals being:

- Avoiding the need to restart vscode.
- Shifting filter list management from the user to the extension.
- Allowing the creation and management of filter list bundles on a per-project basis.

To achieve these goals, I had to split the UX into two parts, a method that is also used in other extensions. However, in the current PR, the project structure is mandatory, which may have led to some usability issues and might not have been fully tested. I'll work on improving this aspect.

If you have any better ideas regarding the UX and workflow, I'd love to hear them.

JeanTracker commented:
I’ve structured the UX workflow to clearly differentiate based on the user’s intended use, as follows:

### User Classification

I have separated the interface for basic users and advanced users.

- Basic Users: These are users who simply set up filter groups and use them for log analysis.
- Advanced Users: These users save filter groups on a project basis and use them to analyze various types of logs.


### UI Structure

The UI is divided into two main components:

- Primary SideBar (Filters): This is intented for basic users.
- ActivityBar (Log Analysis Project): This is specifically for advanced users.

### Basic User Guide

Basic users can add filter groups necessary for log analysis and configure filters within those groups. If they want to reuse a configuration, they can save it, ensuring that the filter setup remains intact even after restarting vscode. If the user has no intention of using advanced features, it is recommended to disable the ActivityBar menu.

### Advanced User Guide

Advanced users should understand that a `project` is a unit that comprises filter groups set up in the SideBar. If multiple projects are needed, the ActivityBar menu can be utilized.

In this menu, users can add or remove projects and select the appropriate project for log analysis, which will then adjust the filter configuration accordingly. Upon selecting a project, users are redirected to the SideBar where they can see the name of the selected project.

Advanced users can create a new project in the ActivityBar, then switch to the SideBar to configure filters. If this process seems cumbersome, there is an option in the ActivityBar to edit filters by modifying a JSON file through the settings. (In this case, a refresh is required for changes to take effect.)

370417 commented:
Thanks for explaining your intentions and design decisions behind this feature.

We tried your solution again. This time, we got it to work, and we appreciate the work you put into improve the capabilities of the extension.

> This led to the introduction of the project feature, with the main goals being:
> - Avoiding the need to restart vscode.

Agreed.

> - Shifting filter list management from the user to the extension.
> - Allowing the creation and management of filter list bundles on a per-project basis.

We agree that:

- Persistent storage should be easier
- Need a solution to switch filters when you switch projects

However, we'd prefer a different solution. Instead of persisting filters into a central config file, we propose to use one config file per project root folder, similar to Prettier's `.prettierrc`. Filters would be autosaved into a `.log-analysis-filters.json` file. This file would be automatically created at the root of the workspace folder when the user creates their first group. If the file already exists, the extension would load its contents into the UI.

Our reasoning for preferring local configuration over central configuration:

- Principle of least surprise: it better follows existing conventions set by other development tools.
- When opening a different project directory, saved filters are automatically loaded. There's no need to manually select a new set of filters.
- Overall UI complexity is lower.
- It avoids potential ui sync issues between projects panel and filter panel (we encountered such a sync issue, but we think it might just be a fixable bug).

Disadvantages of our proposed solution:

- In multi-root workspaces, our solution would present as two sets of filters side by side, similar to the explorer ui. In contrast, a central config would allow the user to pick one set of filters to use.
- In one directory, there is only one configuration. With central configuration, users can switch between different sets of filters within the same directory. However, users are able to achieve the same affect by creating multiple groups, thanks to the groups feature that you added.

We would like to know what you think of the tradeoffs between the two solutions.

JeanTracker commented:
Thank you for sharing another solution proposal and its pros and cons. It's great to see that you're highly focused on user experience (UX), and I'm glad that this extension is being discussed in a way that will further improve it.

Please review the following summarized points and share your feedback.

## Configuration File Management

As you mentioned, several extensions like Prettier support a local configuration approach. However, I believe this may not be suitable for Log Analysis, and I'd like to explain why by comparing it with Prettier.

### Prettier

Prettier has unique code formatter settings for each project, making it essential to support local configuration. Here, the 'project' refers to a specific folder containing source code. This approach helps users maintain consistent code styles across various projects and is an essential feature that considers user convenience.

### Log Analysis

On the other hand, the concept of project in this PR does not refer to a physical structure like a folder. Since the method of log analysis differs depending on the platform or framework, this extension is designed to manage these differences on a 'project' basis. This allows users to utilize VSCode in a way that meets various log analysis needs.

#### Methods for Accessing Logs in VSCode

* Creating a new file and pasting the log to it
* Opening a log file directly
* Dragging and dropping a log file from the file explorer

Log files are independent of the workspace (folder) and can be easily accessed from anywhere. Due to this characteristic, it's important to manage the filter settings for log analysis centrally and consistently. This helps users perform log analysis tasks without confusion.

## Project Updates

I intentionally excluded the feature that automatically updates the filter configuration when the user configures it. I believe it's important to save the user's analysis method needed for log analysis, rather than simply backing up the last log analysis situation. The types of logs we need to analyze are very diverse. It is crucial to specify important analysis methods among them, which requires a clear intent from the user. Additionally, there needs to be a way to analyze logs that do not use the saved analysis methods, allowing the user to freely configure filters for specific log analysis without affecting the existing saved methods. Advanced users can configure multiple projects to record various analysis methods and choose to reconfigure them, but for basic users, this could be a critical issue.

## Project Selection

As mentioned previously, configuring multiple projects is a feature intended for advanced users, while basic users will have a single project selected automatically. Advanced users, having configured projects out of necessity, will understand this well and will not be averse to the process of selecting projects.

SoySauceFor3 commented:
Thanks for the new input. 
We think that your points made for use cases where the tool used for logs without opening a folder makes sense. Therefore we will approve this PR after a few small issues get resolved and the code review is finished. We will do the code review after this comment. 

1. Allow users to add groups without selecting a project. 
![Screenshot 2024-09-03 at 2 08 39 PM](https://github.com/user-attachments/assets/8c3b14c9-456e-4048-bfdd-da5697e63576)
As you said, basic users should not need to be aware of project management, so they should be able to add a group in the FILTER panel without doing anything about projects. However, now according to our test, as you can see in the screenshot, no groups can be created if no project is selected. 
2. Open settings in the Project bar does not work: 
![Screenshot 2024-09-03 at 2 14 11 PM](https://github.com/user-attachments/assets/e39c48a0-7340-45e3-a7f8-977462d46c84)
We assume click this settings button will open the settings json file but it does not. 
3. Save group to projects should NOT allow the "renaming" 
![Screenshot 2024-09-03 at 2 17 10 PM](https://github.com/user-attachments/assets/9579df9c-0088-4878-9620-c0c49e42ea9a)
As the screenshot shows, once click the save button in the filter panel, there is a text input to collect the project name. However, first the program already know which project is selected; second, this text input does renaming to the project ---- which can lead to surprise and confusion. So we think it is best to just remove this text input and automatically save the changes to the selected project. If users want to do renaming, they can just go to the project bar and do the renaming there. 
4. (minor) the refresh mechanism 
We don't know if it is feasible, but we would like to see more auto fresh than now. For example: if a user renames the selected project, we would like to see that WITHOUT clicking the refresh button, the user can see the name of the project updated when coming back to the filter panel. Right now the name does not get updated automatically. 
The root cause solution is probably watching changes to the config file. But if it is not feasible, a partial fix to the rename case is also appreciated. 

While we work on these issues, we will also start code review. Thanks for the work you put in, we really appreciate it.

JeanTracker commented:
Thank you for understanding the proposal well.

1) Project Selection
Could you check the commit you used for testing again? The last commit is 20581d46ccff8e0cc23427a50219ce369d03cda7. When testing with this commit, the project name (NONAME) should appear next to the `FILTERS` tab as shown in the image below.
![image](https://github.com/user-attachments/assets/4a89dc65-3af4-406a-abfb-f4f54e9fd777)
When only one project is configured, it is automatically selected. If an advanced user has configured two or more projects, they will need to select the desired project.

2) Settings
Could you share your development environment? I am testing on a MacBook. Please check if the home directory is being accessed correctly using [this code](https://github.com/SoySauceFor3/log-analysis/commit/20581d46ccff8e0cc23427a50219ce369d03cda7#diff-bb57a3bd912abc3ec2e729cb8a743838487677a5517683d0c8913a3619ac296aR11).

3) Project Name Change
When changing the project name, it is automatically saved and updated in the UX, reflecting the change to the user. There are two ways to change the project name:
- Click the save icon in the `FILTERS` tab.
- Click the project name edit icon in the `LOG ANALYSIS: PROJECT` menu.

4) Auto-Update Policy
As mentioned earlier, in the case of automatic updates, the changes are immediately applied, and the user can confirm them. For manual updates, the refresh icon must be clicked for the changes in the configuration file to take effect. Whether to update the project filters is up to the user. Users should be able to reconfigure filters freely for one-time use to analyze logs, while also having the option to restore a backed-up project. The refresh button exists for this purpose.

**Auto Update**
- Project filter updates (`FILTERS` tab > Save icon)
- Project name changes (`LOG ANALYSIS: PROJECT` menu > Project name edit icon)
- Select Project (`LOG ANALYSIS: PROJECT` menu > Select a project)

**Manual Update**
- Editing directly the configuration file via an editor (`LOG ANALYSIS: PROJECT` menu > Settings icon)
- Reconfiguring the project filters without updating the project

I share this additional explanation along with answers for each test. Please provide your feedback.

JeanTracker commented:
Thanks for testing again.

1. Acting like Google Docs is a very good idea. Basic users will initially use the `NONAME` project, and even if they rename it, the same project remains active. For advanced users who have created multiple projects, I have modified it so that when no project is selected, a blank project named `NONAME` is selected by default.

2. In the screenshot, the files are open in the open editors but not showing in the editor window. Could this be a bug in VSCode? :)

3. In the `FILTERS` tab, I have modified it so that only the configured filters are saved, without changing the project name. The status bar now informs the user whether the project has been saved.

4. There was a bug. It was reproduced, and I have fixed it.

SoySauceFor3 commented on src/settings.ts:
```diff
@@ -0,0 +1,81 @@
+import { State } from "./extension";
+import { Project, Group, Filter, generateSvgUri } from './utils';
+
+const settingFilename: string = path.join(os.homedir(), 'vscode_log_analysis.json');
```
store home dir can be messy to users --- the file exists after they uninstall the extension or vscode. 

I recommend context.globalStorageUri (context var from the activate function)

JeanTracker replied:
That's good idea. thanks.

SoySauceFor3 replied:
Saw the change. I understand the need to store the Uri to some "global vairables" like state, but I don't think it is a good idea to pass state to "getSettingFile", "openSettings", "readSettings". It is better to just pass the Uri itself, which will make the state management less coupled.

JeanTracker replied:
I fixed it.

SoySauceFor3 commented on src/settings.ts:
```diff
@@ -0,0 +1,81 @@
+
+const settingFilename: string = path.join(os.homedir(), 'vscode_log_analysis.json');
+
+export function openSettings(state: State) {
```
state is not used

JeanTracker replied:
I fixed it.

SoySauceFor3 replied:
Same comment as above on not passing state var around.

JeanTracker replied:
I fixed it.

SoySauceFor3 commented on src/settings.ts:
```diff
@@ -0,0 +1,81 @@
+        } catch (e) {
+            vscode.window.showErrorMessage('The settings file is broken');
+        }
+    } else {
+        saveSettings(projects);
+    }
```
a write in a Read function is unexpected. Do you want to make it more explicit by either add comments or move the corner case to the caller?

JeanTracker replied:
I moved this to `refreshSettings` for create new setting file.

SoySauceFor3 commented on src/utils.ts:
```diff
@@ -35,3 +42,7 @@ export function generateSvgUri(
   return vscode.Uri.parse(dataUri);
 }
+
+export function setStatusBarMessage(message: string) {
+  vscode.window.setStatusBarMessage(`LOG ANALYSIS: ${message}`, 5000);
+}
```
Do you want to use this more? Now it is just used 1 time.

JeanTracker replied:
I know but this will be useful in the future.

SoySauceFor3 replied:
got it

SoySauceFor3 reviewed (changes requested):
Thanks for the hard work 💯 
If you can, pls make the changes a separate commit so when review I can see just the new diff.

JeanTracker commented:
> Thanks for the hard work 💯 If you can, pls make the changes a separate commit so when review I can see just the new diff.

I understand that it may be inconvenient to check the differences, but I kindly ask for your understanding as splitting commits for code reviews is not something we generally do.

JeanTracker commented on src/commands.ts:
```diff
@@ -288,11 +211,22 @@ export function refreshEditors(state: State) {
+  console.log("update project tree view");
+  state.projectTreeViewProvider.update(state.projects);
+}
+
```
I removed it.

JeanTracker commented on src/commands.ts:
```diff
@@ -288,11 +211,22 @@ export function refreshEditors(state: State) {
   state.filterTreeViewProvider.refresh();
 }
 
+export function updateFilterAndTreeView(state: State) {
```
I fixed it.

JeanTracker commented on src/commands.ts:
```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
+    saveSettings(state.projects);
+    updateProjectAndTreeView(state);
+  }
+}
```
It seems appropriate for the extension to handle setting the title of the view created by the extension.

`updateExplorerTitle` is reflected after the following commands:
- deleteProject
- editProject
- refreshSettings
- selectProject

SoySauceFor3 replied:
It seems that you are confused on what we are trying to convey. What we want is that instead of the current callsite of `updateExplorerTitle`, we want functions like "deleteProject" to call `updateExplorerTitle`.

JeanTracker replied:
It seems more efficient for readability to handle changing the explorer view's title separately. even when the user executes `selectProject`, it would be more natural and efficient in terms of readability to call  `vscode.commands.executeCommand('workbench.view.explorer')` from extension.ts. 

The `updateExplorerTitle` function has been moved to commands.ts.

JeanTracker commented on src/commands.ts:
```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
+
+    const projectLength = state.projects.push(project);
+    state.selectedIndex = projectLength - 1;
+  }
```
There is no duplication because if there are no projects, a `NONAME` project is created by reading the configuration file. I have modified it so that after creating the `NONAME` project, it is saved to the file.

SoySauceFor3 replied:
Saw the new commit, fine with me.

JeanTracker commented on src/commands.ts:
```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
   refreshEditors(state);
 }
+
+export function projectSelected(treeItem: vscode.TreeItem, state: State): boolean {
```
I fixed it.

JeanTracker commented on src/commands.ts:
```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
+        });
+      });
+    });
+    state.selectedIndex = selectedIndex;
+
+    const project = state.projects[selectedIndex];
+    project.selected = true;
```
`project.selected` is information intended for use in project items, and it is not duplicated.

SoySauceFor3 replied:
I saw how selected property is used in ProjectItem's constructor, but in the state object, the selectedIndex and projects' selected property are duplicated information, which can go out of sync and create bugs. 

I will suggest maybe remove the "selectedIndex" var and when needing to know/use this, just compute on state.projects directly.

JeanTracker replied:
I removed `state.selectedIndex`.

JeanTracker commented on src/extension.ts:
```diff
@@ -4,40 +4,64 @@ import {
+    focusProvider: new FocusProvider(groups),
   };
+
+  refreshSettings(state);
```
In the case of `refreshSettings`, it configures the state. Its role is to populate the data before the view is constructed. On the other hand, `updateExplorerTitle` updates after the view is created.

JeanTracker commented on src/extension.ts:
```diff
@@ -75,17 +106,68 @@ export function activate(context: vscode.ExtensionContext) {
+        return;
+      }
+      if (projectSelected(treeItem, state)) {
+        view.title = 'Filters (' + treeItem.label + ')';
```
I fixed it.

JeanTracker commented on src/extension.ts:
```diff
@@ -179,7 +261,13 @@ export function activate(context: vscode.ExtensionContext) {
+      if (state.selectedIndex === -1) {
+        vscode.window.showInformationMessage('You need to add a project first!');
+        return;
+      }
```
I fixed it.

JeanTracker commented on src/filterTreeViewProvider.ts:
```diff
@@ -25,19 +25,34 @@ export class FilterTreeViewProvider implements vscode.TreeDataProvider<vscode.Tr
-  refresh(): void {
-    console.log("in refresh");
-    this._onDidChangeTreeData.fire(undefined);
+  refresh(element?: vscode.TreeItem): void {
+    if (element === undefined) {
+      console.log("refresh all");
+    } else {
+      console.log("refresh item");
+    }
+    this._onDidChangeTreeData.fire(element);
+  }
```
This content has been added as part of the necessary improvements for performance optimization. It will be utilized in a future PR.

JeanTracker commented on src/filterTreeViewProvider.ts:
```diff
@@ -72,6 +87,11 @@ export class GroupItem extends vscode.TreeItem {
+    this.update(filter);
+  }
+
+  update(filter: Filter) {
```
This content has been added as part of the necessary improvements for performance optimization. It will be utilized in a future PR.

JeanTracker commented on src/filterTreeViewProvider.ts:
```diff
@@ -25,19 +25,34 @@ export class FilterTreeViewProvider implements vscode.TreeDataProvider<vscode.Tr
+    this.update(group);
+  }
+
+  update(group: Group) {
```
This content has been added as part of the necessary improvements for performance optimization. It will be utilized in a future PR.

JeanTracker commented on src/focusProvider.ts:
```diff
@@ -45,4 +45,8 @@ export class FocusProvider implements vscode.TextDocumentContentProvider {
   }
+
+  update(groups: Group[]) {
+    this.groups = groups;
```
The group changes when the project is switched. It is necessary to update the reference value passed when the provider is created.

SoySauceFor3 replied:
I don't have any question on "this.groups = groups;" (seems that your response is about this statement). I want to know why you don't call "this.refresh()" here for focusProvider.
